src/runtime/HalideBuffer.h

```diff
@@ -2029,35 +2016,46 @@ class Buffer {
 
         // flatten dimensions where possible to make a larger inner
         // loop for autovectorization.
-        int d = dimensions();
+        int d = dimensions;
         for (int i = 1; i < d; i++) {
             bool flat = true;
             for (int j = 0; j < N; j++) {
                 flat = flat && t[i-1].stride[j] * t[i-1].extent == t[i].stride[j];
             }
             if (flat) {
                 t[i-1].extent *= t[i].extent;
-                for (int j = i; j < dimensions(); j++) {
+                for (int j = i; j < d; j++) {
                     t[j] = t[j+1];
                 }
                 i--;
                 d--;
+                t[d].extent = 1;
             }
         }
 
-        bool innermost_strides_are_one = false;
-        if (dimensions() > 0) {
-            innermost_strides_are_one = true;
-            for (int j = 0; j < N; j++) {
-                innermost_strides_are_one &= t[0].stride[j] == 1;
+        bool innermost_strides_are_one = true;
+        if (dimensions > 0) {
+            for (int i = 0; i < N; i++) {
+                innermost_strides_are_one &= (t[0].stride[i] == 1);
             }
         }
 
-        if (innermost_strides_are_one) {
-            for_each_value_helper<true>(f, dimensions() - 1, t, begin(), (other_buffers.begin())...);
-        } else {
-            for_each_value_helper<false>(f, dimensions() - 1, t, begin(), (other_buffers.begin())...);
-        }
+        return innermost_strides_are_one;
+    }
+
+    template<typename Fn, typename ...Args, int N = sizeof...(Args) + 1>
+    void for_each_value_impl(Fn &&f, Args&&... other_buffers) const {
+        Buffer<>::for_each_value_task_dim<N> *t =
+        (Buffer<>::for_each_value_task_dim<N> *)HALIDE_ALLOCA((dimensions()+1) * sizeof(for_each_value_task_dim<N>));
+        // Move the preparatory code into a non-templated helper to
+        // save code size.
+        const halide_buffer_t *buffers[] = {&buf, (&other_buffers.buf)...};
+        bool innermost_strides_are_one = Buffer<>::for_each_value_prep(t, buffers);
+
+        Buffer<>::for_each_value_helper(f, dimensions() - 1,
```

Comment:
Note that because we enter `Buffer<>` at this point, all the support code (advance_ptrs, etc), is only called from within Buffer<>, so that it doesn't get a separate instantiation per T, D, parameter.